README.md
# Spark Lightning Wallet

[![npm release](https://img.shields.io/npm/v/spark-wallet.svg)](https://www.npmjs.com/package/spark-wallet)
[![build status](https://api.travis-ci.org/shesek/spark-wallet.svg)](https://travis-ci.org/shesek/spark-wallet)
[![docker pulls](https://img.shields.io/docker/pulls/shesek/spark-wallet.svg)](https://hub.docker.com/r/shesek/spark-wallet)
[![npm installs](https://img.shields.io/npm/dt/spark-wallet?label=npm%20installs)](https://www.npmjs.com/package/spark-wallet)
[![downloads](https://img.shields.io/github/downloads/shesek/spark-wallet/total.svg?color=blueviolet)](https://github.com/shesek/spark-wallet/releases)
[![MIT license](https://img.shields.io/github/license/shesek/spark-wallet.svg)](https://github.com/shesek/spark-wallet/blob/master/LICENSE)
[![Pull Requests Welcome](https://img.shields.io/badge/PRs-welcome-brightgreen.svg)](http://makeapullrequest.com)
[![IRC](https://img.shields.io/badge/chat-on%20freenode-brightgreen.svg)](https://webchat.freenode.net/?channels=spark-wallet)

:zap: Simple & minimalistic
:zap: Purely off-chain
:zap: Progressive Web App
:zap: Personalizable themes
:zap: Mobile and desktop apps
:zap: Automatic self-signed TLS
:zap: LetsEncrypt integration
:zap: Tor hidden service (v3)
:zap:

![Spark screenshot](doc/img/spark-header.png)


<img align="right" width="25%" src="doc/img/payment.gif">

## Contents

- [Introduction](#introduction)
- [Server installation](#server-installation) (for web access)
- [Desktop apps](#desktop-apps) (for Linux, macOS and Windows)
- [Mobile app](#mobile-app) (for Android, iOS coming soon)
- [Progressive Web App](#progressive-web-app)
- [GUI settings & controls](#gui-settings--controls)
- [Browser support](#browser-support)
- [Developing](#developing)
- [Code signing & reproducible builds](#code-signing--reproducible-builds)
- [CLI options](#cli-options)

## Introduction

Spark is a minimalistic wallet GUI for c-lightning, accessible over the web or
through mobile and desktop apps (for Android, Linux, macOS and Windows).
It is currently oriented for technically advanced users and is not an all-in-one package,
but rather a "remote control" interface for a c-lightning node that has to be managed separately.

Sparks supports sending and receiving payments, viewing history, and managing channels.

Spark is a purely off-chain wallet, with no on-chain payments.
This allows Spark to fully realize the awesome UX enabled by lightning,
without worrying about the complications and friction of on-chain.
This might change someday.

Spark has a responsive UI suitable for mobile, tablet and desktop devices,
but is best optimized for use on mobile.

> ⚠️ Spark is beta-quality software under active development, please use with care.

Big shout out to [Blockstream](https://blockstream.com) for generously sponsoring this work!

## Server installation

Requires a running [c-lightning](https://github.com/ElementsProject/lightning) node
(see setup instructions in [the official docs](https://github.com/ElementsProject/lightning/blob/master/doc/INSTALL.md)
or [this tutorial](https://hackernoon.com/harnessing-lightning-for-instant-bitcoin-transacting-a-tutorial-3b9dcdc00552))
and nodejs v6.0 or newer (nodejs v8 is recommended, see [instructions here](https://nodejs.org/en/download/package-manager/).
If you're running into permission issues,
[try this](https://docs.npmjs.com/getting-started/fixing-npm-permissions#option-two-change-npms-default-directory).)

```bash
$ npm install -g spark-wallet

$ spark-wallet # defaults: --ln-path ~/.lightning --port 9737
```

Or simply `$ npx spark-wallet`, which will install and start Spark in one go.

Spark will generate and print a random username and password that'll be used to login into the wallet,
and persist them to `~/.spark-wallet/cookie` (can be controlled with `--cookie-file <path>`).
To specify your own login credentials, set `--login [user]:[pass]` or the `LOGIN` environment variable.

To access the wallet, open `http://localhost:9737/` in your browser and login with the username/password.

You may also start Spark with `--pairing-url`, which will print a URL with an embedded access token,
which you can open in your browser to login into the wallet without using the username/password.
`--pairing-qr` provides the same URL as a QR (useful for mobile pairing).

See `$ spark-wallet --help` for the full list of command-line options (also available under [*CLI options*](#cli-options)).

#### Configuration file

Spark reads configuration options from `~/.spark-wallet/config` (can be overridden with `--config/-c <path>`).
The expected format is one `key=value` per line, like so:

```ini
ln-path=/data/lightning/testnet
login=bob:superSecretPassword
port=8000
```

#### Connecting remotely

To accept remote connections, set `--host <listen-address>` (shorthand `-i`, e.g. `-i 0.0.0.0`).
This will automatically enable TLS with a self-signed certificate.

For more information on TLS, instructions for setting up a CA-signed certificate using the built-in LetsEncrypt integration
and for adding the self-signed certificate to Android,
see [`doc/tls.md`](doc/tls.md).

To start Spark as a Tor hidden service, set `--onion`. You don't need Tor pre-installed for this to work.
See [`doc/onion.md`](doc/onion.md) for more details,
the advantages of using an hidden service, and instructions for connecting from Android.
*This is highly recommended.*

#### Deploy with Docker

Spark is also available as a Docker image that comes bundled with bitcoind and c-lightning.
See [`doc/docker.md`](doc/docker.md) for details.

#### Adding to startup with `systemd`

See [`doc/startup-systemd.md`](doc/startup-systemd.md).

## Desktop apps

Electron-based desktop apps for Linux (packaged as `deb`, `AppImage`, `snap` and `tar.gz`),
macOS (as `zip`) and Windows (installer and a portable) are available for download from the
[releases page](https://github.com/shesek/spark-wallet/releases).

The desktop apps comes bundled with the Spark server-side component. If you're connecting to a local
c-lightning instance, you [can configure](https://user-images.githubusercontent.com/877904/44618385-68f40100-a87d-11e8-891b-79a0f2d59751.png)
the desktop app to connect to it directly without manually setting up the Spark server.

Connecting to a remote c-lightning instance requires setting up the Spark server on the same machine
running c-lightning and connecting through it.

## Mobile app

A Cordova-based native app for Android is available for download from the
[Google Play app store](https://play.google.com/store/apps/details?id=com.spark.wallet)
or from the [releases page](https://github.com/shesek/spark-wallet/releases).

The app requires a Spark server to communicate with, which you need [to setup](#server-installation) as a prerequisite.

When the app starts for the first time, you'll need to configure your Spark server URL and API access key.
You can print your access key to the console by starting Spark with `--print-key/-k`.
You can also scan this information from a QR, which you can get with `--pairing-qr/-Q`.

For the native app to properly communicate with the server, the TLS certificate has to be
[signed by a CA](doc/tls.md#letsencrypt-integration)
or manually [added as a user trusted certificate](doc/tls.md#add-as-trusted-certificate-to-android).

## Progressive Web App

You can install Spark as a [PWA](https://developer.mozilla.org/en-US/Apps/Progressive) (on mobile and desktop) to get a more native-app-like experience,
including an home launcher that opens up in full screen, system notifications and faster load times.

Available in Chrome mobile under `⋮` -> `Add to homescreen` ([see here](https://imgur.com/zVe1sOH)),
in Chrome desktop under `More tools` -> `Install to desktop` ([see here](https://i.imgur.com/Pj6FpGA.png))
and in Firefox mobile with an icon next to the address bar ([see here](https://mdn.mozillademos.org/files/15762/add-to-home-screen-icon.png)).

Installing the PWA requires TLS and a [CA-signed certificate](doc/tls.md#letsencrypt-integration)
(unless accessed via `localhost`).

Compared to the PWA, the main advantages of the mobile and desktop apps are
the ability to handle `lightning:` URIs,
better security sandbox (detached from the browser)
and static client-side code.

**Note for iOS users:** iOS [does](https://github.com/webrtc/samples/issues/933) [not](https://stackoverflow.com/questions/46228218/how-to-access-camera-on-ios11-home-screen-web-app/46350136)
allow PWAs to use WebRTC (required for the QR scanner), but it works otherwise.
The QR scanner works if you access Spark without using the PWA "Add to homescreen" feature.

<img align="right" width="33%" src="doc/img/gui-controls.gif">

## GUI settings & controls

- **Pay** and **Request** are pretty intuitive and don't require much explaining. Try them!

- **Display unit:** Click the balance on the top-right or the unit in the "request payment" page to toggle the currency display unit.
  The available options are sat, bits, milli, btc and usd.

- **Theme switcher:** Click the theme name on the bottom-right to change themes (you can choose between 16 [bootswatch](https://bootswatch.com) themes).

- **Payment details:** Click on payments in the list to display more details.
  (note that the fee shown includes c-lightning's [overpayment randomization](https://github.com/ElementsProject/lightning/issues/1089))

- **Expert mode:** Click the version number on the bottom-left to toggle expert mode.
  This will add two new menu items, "*Logs*" and
  ["*RPC Console*"](https://user-images.githubusercontent.com/877904/44295439-7fb0bb80-a2b1-11e8-8506-f5afb1c9f1d7.png),
  and display yaml dumps with additional information throughout the app.

- **Node address:** Click the node id on the footer to open the node info page which displays your node address (as text and QR).

- **Channel management:** Click the "*Channels*" button inside the node info page to show and manage channels.

## Browser support

Supported on recent desktop and mobile version of Chrome, Firefox and Safari.
IE is unsupported.

Requires iOS 11.2+ for WebRTC (used by the QR scanner), but works otherwise with iOS 9+.
Chrome on iOS does not support WebRTC.

## Developing

Spark is written in a reactive-functional style using [rxjs](https://github.com/ReactiveX/rxjs) and [cycle.js](https://cycle.js.org),
with bootstrap for theming and a nodejs/express server as the backend.

To start a development server with live compilation for babel, browserify, pug and stylus, run:

```bash
$ git clone https://github.com/shesek/spark-wallet && cd spark-wallet
$ npm install
$ npm start -- --ln-path /data/lightning
```

Spark can be built from source using the following commands (more efficient than running the live compliation development server):

```bash
$ git clone https://github.com/shesek/spark-wallet && cd spark-wallet
$ npm install
$ npm run dist:npm
$ ./dist/cli.js --ln-path /data/lightning
```

Cordova builds can be prepared with `npm run dist:cordova`.
The `.apk` file will be available in `cordova/platforms/android/app/build/outputs/apk/debug/`.

Electron builds can be prepared with `npm run dist:electron`.
They will be available under `electron/dist`.

To get more verbose server-side logging, start the server with `--verbose` (or `-V`).

To get more verbose output in the browser developer console, set `localStorage.debug = 'spark:*'`.

See [`doc/dev-regtest-env.md`](doc/dev-regtest-env.md) for instructions setting up a regtest environment with multiple wallets.

Pull requests, suggestions and comments are welcome!

## Code signing & reproducible builds

Signed distribution checksums are available in the git repo at
[`SHA256SUMS.asc`](SHA256SUMS.asc)
(updated with every versioned release)
and on the [releases page](https://github.com/shesek/spark-wallet/releases).
Git version tags are signed too.

The releases are signed by Nadav Ivgi (@shesek).
The public key can be verified on [keybase](https://keybase.io/nadav),
[github](https://api.github.com/users/shesek/gpg_keys),
[twitter](https://twitter.com/shesek) (under bio),
[HN](https://news.ycombinator.com/user?id=nadaviv)
or on [a domain he's known to control](https://www.bitrated.com/.keys/staff/nadav.asc).

To install the signed NPM package, download it from
[the releases page](https://github.com/shesek/spark-wallet/releases), verify the hash
and install using `$ npm install -g spark-wallet-[x.y.z]-npm.tgz`.

To install the signed Docker image, get the image hash from `SHA256SUMS.asc` and install it with
`$ docker pull shesek/spark-wallet@sha256:[image-hash-verified-by-be-signed]`.

The NPM package, Linux `AppImage`/`snap`/`tar.gz` builds, macOS `zip` builds and the Windows installer
[are deterministically reproducible](doc/reproducible-builds.md).

## CLI options

```bash
$ spark-wallet --help

  A minimalistic wallet GUI for c-lightning

  Usage
    $ spark-wallet [options]

  Options
    -l, --ln-path <path>     path to c-lightning data directory [default: ~/.lightning]
    -p, --port <port>        http(s) server port [default: 9737]
    -i, --host <host>        http(s) server listen address [default: localhost]
    -u, --login <userpwd>    http basic auth login, "username:password" format [default: generate random]
    -C, --cookie-file <path> persist generated login credentials to <path> or load them [default: ~/.spark-wallet/cookie]
    --no-cookie-file         disable cookie file [default: false]

    --rate-provider <name>   exchange rate provider, one of "bitstamp" or "wasabi" (requires tor) [default: bitstamp]
    --no-rates               disable exchange rate lookup [default: false]
    --proxy <uri>            set a proxy for looking up rates, e.g. socks5h://127.0.0.1:9050 [default: none]

    --force-tls              enable TLS even when binding on localhost [default: enable for non-localhost only]
    --no-tls                 disable TLS for non-localhost hosts [default: false]
    --tls-path <path>        directory to read/store key.pem and cert.pem for TLS [default: ~/.spark-wallet/tls/]
    --tls-name <name>        common name for the TLS cert [default: {host}]

    --letsencrypt <email>    enable CA-signed certificate via LetsEncrypt [default: false]
    --le-port <port>         port to bind LetsEncrypt verification server [default: 80]
    --le-noverify            skip starting the LetsEncrypt verification server [default: start when {letsencrypt} is set]
    --le-debug               display additional debug information for LetsEncrypt [default: false]

    -o, --onion              start Tor Hidden Service (v3) [default: false]
    -O, --onion-path <path>  directory to read/store hidden service data [default: ~/.spark-wallet/tor/]
    --onion-nonanonymous     setup hidden service in non-anonymous mode [default: false]

    -k, --print-key          print access key to console (for use with the Cordova/Electron apps) [default: false]
    -q, --print-qr           print QR code with the server URL [default: false]
    -Q, --pairing-qr         print QR code with embedded access key [default: false]
    -P, --pairing-url        print URL with embedded access key [default: false]
    --public-url <url>       override public URL used for QR codes [default: http(s)://{host}/]

    --allow-cors <origin>    allow browser CORS requests from <origin> (USE WITH CARE) [default: off]
    --no-webui               run API server without serving client assets [default: false]
    --no-test-conn           skip testing access to c-lightning rpc (useful for init scripts) [default: false]

    -c, --config <path>      path to config file [default: ~/.spark-wallet/config]
    -V, --verbose            display debugging information [default: false]
    -h, --help               output usage information
    -v, --version            output version number

  Example
    $ spark-wallet -l ~/.lightning

  All options may also be specified as environment variables:
    $ LN_PATH=/data/lightning PORT=8070 NO_TLS=1 spark-wallet
```

## License

[MIT](LICENSE)


CHANGELOG.md
# Changelog

## 0.3.1 - 2021-11-08

- Cordova: Target Android SDK version 30

- Docker: Update to Bitcoin Core v22.0 and c-lightning v0.10.2

## 0.3.0 - 2021-11-03

- BOLT12 offers! ✨🎉

  - Send payments using BOLT12 offers (including support for `any` amounts, quantities, fiat denomination and sending a `payer_note` to the payee).

  - Create reusable BOLT12 offers to receive multiple payments (basic support only).

  - Pull payments using BOLT12 withdrawal offers (`send_invoice=true`).

  - Recurrence is not supported yet -- to be added in a followup release.

  **Requires c-lightning v0.10.1+** with `experimental-offers` enabled.

- UI: Non-blocking payments with `paystatus` tracking

  Payments are now sent in the background without blocking the UI,
  with pending/failed payments shown in the homepage payment history.

- UI: Improve the main screen initial loading times

  Pay/Request can now be used immediately, without waiting for the wallet's data to load first.

- UI: Always display payment amounts in USD terms too

- UI: Improved error reporting and display

- Docker: Allow (and recommend) mounting the directory with the `lightning-rpc` socket file
  (fixes a [compatibility issue with BTCPay](https://github.com/btcpayserver/btcpayserver-docker/pull/462))

- Fix: Properly display the NoScript message on Firefox (#183)

- Update clightning-client to v0.1.4

- Use `newaddr all` to avoid creating new addresses when switching types

  This raises the minimally supported c-lightning version to v0.7.1.

- docker: Update to Debian 11, Bitcoin Core v0.21.0, c-lightning v0.10.1 and nodejs v16.8

- Reproducibility: Due to changes in third-party dependencies, the Android `apk` build and the Windows portable runner are no longer reproducible, while the Linux `AppImage` now is. See [more details here](doc/reproducible-builds.md).

## 0.2.17 - 2020-09-03

Hotfix release - Fix YAML rendering of payments with unknown destination/description. This bug only manifests when expert mode is enabled.

## 0.2.16 - 2020-08-11

Hotfix release

- Properly handle outgoing payments without an associated requested amount (5c0a61ec40eb)

- Workaround for incorrect bolt11 string reported by c-lightning (#151, https://github.com/ElementsProject/lightning/issues/3921)

- Fix the build script to properly install the client devDependencies (#152)

## 0.2.15 - 2020-08-07

- Support multi-part payments, grouped together in the UI (#147)

- Extract destination and description of outgoing payments from the BOLT11 request

- Fix bug in automatic credentials generation (#145)

## 0.2.14 - 2020-04-22

- Fix: Resolve desktop app regression introduced in the last release (#143)

- Electron: Update to v8

- Docker: Updade nodejs to v12

## 0.2.13 - 2020-04-11

- Fix: Resolve TLS issues with NodeJS 10 (#55, h/t @sumBTC!)

- Fix: Some QR codes not read properly in the web QR scanner (#134)

- Docker: Update Bitcoin Core to v0.19.1 and c-lightning to v0.8.1

## 0.2.12 - 2019-12-22

- Docker: Update to c-lightning v0.8.0

- Docker: Compatibility with c-lightning v0.8.0 network subdirectories

- Detect if the "base directory" is provided and default to the bitcoin mainnet network subdirectory within in. (#130)

  This allows users to start spark using the existing command they use (with i.e. `--ln-path ~/.lightning`)
  to ease the upgrade to c-lightning v0.8.0, but this usage is considered deprecated and will eventually be removed.
  Users are advised to explicitly specify the path to the network subdirectory with `--ln-path ~/.lightning/<network>`.

- Don't display unconfirmed onchain balance (#129)

- Cordova Android: Allow connecting to server in cleartext (#132)

  This was possible before v0.2.11 which updated the Android SDK to version 28.
  Not really recommended, but useful for development and testnet/regtest demonstrations.

## 0.2.11 - 2019-12-15

- Fix channels view (#128)

- Fix Android crashes caused by plugin-local-notifications (https://github.com/katzer/cordova-plugin-local-notifications/issues/1541)

## 0.2.10 - 2019-12-08

- Docker: Update to Bitcoin Core v0.19.0.1 and c-lightning v0.7.3

- Fix bug with missing channel reserve (3d1f3b2cc84627ac38ff64bd09aa1d3b5fd74bfe)

## 0.2.9 - 2019-10-17

- *BREAKING CHANGE:* Dropped support for c-lightning before 0.7.0

- Compatibility with c-lightning v0.7.3 (upcoming, currently rc), as well as previous v0.7.x with allow-deprecated-apis=false

- Docker: Add standalone arm32v7 and arm64v8 images (#93)

  Available using the new multiarch image at `shesek/spark-wallet:standalone` and `shesek/spark-wallet:VERSION-standalone`,
  or directly using `shesek/spark-wallet:standalone-arm32v7` and `shesek/spark-wallet:VERSION-standalone-arm32v7`.

  Thanks @NicolasDorier!

- Docker: Upgrade to Bitcoin Core v0.18.1 and c-lightning v0.7.2

- Upgrade to RxJS v6, Babel v7, Cycle-run v10

- Allow enabling CORS using `--allow-cors <origin>` (#101)

- Some minor UI improvements to the balance overview (soon to be replaced entirely)

## 0.2.8 - 2019-06-09

- Security: Make npm-shrinkwrap actually ship with the npm package ([5827098ec](https://github.com/shesek/spark-wallet/commit/5827098ecc301e449e58e6077b0abf4e63aa9635))

- Fix: Make QR scanner work in dark themes (a regression introduced in v0.2.6, #91)

- Fix: Make the "Open wallet" link (bitcoin:<addr>) work in the Android app

- Docker: Use tini to init the container process (#90, thanks @NicolasDorier)

- Update client-side npm dependencies

## 0.2.7 - 2019-06-07

- Fix a bug in the homepage on-chain/channel balances display

## 0.2.6 - 2019-06-07

- Changed the way the balance shown in the top-right corner is calculated. See:

  https://github.com/shesek/spark-wallet/issues/62

  https://github.com/shesek/spark-wallet/commit/9b500c48c4c32fd61a11995410f17adf9f6cdf8d

- Display channel and on-chain balances separately on the homepage (#67)


src/cli.js
#!/usr/bin/env node

const args = require('meow')(`
    Usage
      $ spark-wallet [options]

    Options
      -l, --ln-path <path>     path to c-lightning data directory [default: ~/.lightning]
      -p, --port <port>        http(s) server port [default: 9737]
      -i, --host <host>        http(s) server listen address [default: localhost]
      -u, --login <userpwd>    http basic auth login, "username:password" format [default: generate random]
      -C, --cookie-file <path> persist generated login credentials to <path> or load them [default: ~/.spark-wallet/cookie]
      --no-cookie-file         disable cookie file [default: false]

      --rate-provider <name>   exchange rate provider, one of "bitstamp" or "wasabi" (requires tor) [default: bitstamp]
      --no-rates               disable exchange rate lookup [default: false]
      --proxy <uri>            set a proxy for looking up rates, e.g. socks5h://127.0.0.1:9050 [default: none]

      --force-tls              enable TLS even when binding on localhost [default: enable for non-localhost only]
      --no-tls                 disable TLS for non-localhost hosts [default: false]
      --tls-path <path>        directory to read/store key.pem and cert.pem for TLS [default: ~/.spark-wallet/tls/]
      --tls-name <name>        common name for the TLS cert [default: {host}]

      --letsencrypt <email>    enable CA-signed certificate via LetsEncrypt [default: false]
      --le-port <port>         port to bind LetsEncrypt verification server [default: 80]
      --le-noverify            skip starting the LetsEncrypt verification server [default: start when {letsencrypt} is set]
      --le-debug               display additional debug information for LetsEncrypt [default: false]

      -o, --onion              start Tor Hidden Service (v3) [default: false]
      -O, --onion-path <path>  directory to read/store hidden service data [default: ~/.spark-wallet/tor/]
      --onion-nonanonymous     setup hidden service in non-anonymous mode [default: false]

      -k, --print-key          print access key to console (for use with the Cordova/Electron apps) [default: false]
      -q, --print-qr           print QR code with the server URL [default: false]
      -Q, --pairing-qr         print QR code with embedded access key [default: false]
      -P, --pairing-url        print URL with embedded access key [default: false]
      --public-url <url>       override public URL used for QR codes [default: http(s)://{host}/]

      --allow-cors <origin>    allow browser CORS requests from <origin> (USE WITH CARE) [default: off]
      --no-webui               run API server without serving client assets [default: false]
      --no-test-conn           skip testing access to c-lightning rpc (useful for init scripts) [default: false]

      -c, --config <path>      path to config file [default: ~/.spark-wallet/config]
      -V, --verbose            display debugging information [default: false]
      -h, --help               output usage information
      -v, --version            output version number

    Example
      $ spark-wallet -l ~/.lightning

    All options may also be specified as environment variables:
      $ LN_PATH=/data/lightning PORT=8070 NO_TLS=1 spark-wallet

`, { flags: { lnPath: {alias:'l'}, login: {alias:'u'}, cookieFile: {alias:'C'}
            , port: {alias:'p'}, host: {alias:'i'}
            , leNoverify: {type:'boolean'}, leDebug: {type:'boolean'}
            , printKey: {type:'boolean', alias:'k'}, printQr: {type:'boolean', alias:'q'}
            , pairingQr: {type:'boolean', alias:'Q'}, pairingUrl: {type:'boolean', alias:'P'}
            , onion: {type:'boolean',alias:'o'}, onionPath: {alias:'O'}, onionNonanonymous: {type:'boolean'}
            , config: {alias:'c'}, verbose: {alias:'V', type:'boolean'}
} }).flags

// Load config file
const os = require('os'), fs = require('fs'), path = require('path'), ini = require('ini')
    , confPath = args.config || process.env.CONFIG || path.join(os.homedir(), '.spark-wallet', 'config')
    , fileConf = fs.existsSync(confPath) ? ini.parse(fs.readFileSync(confPath, 'utf-8')) : {}

const conf = Object.assign({}, fileConf, args)

// Set config options from argv and file as environment variables
Object.keys(conf).filter(k => k.length > 1)
  .map(k => [ k.replace(/([^A-Z_])([A-Z])/g, '$1_$2').replace(/-/g, '_').toUpperCase(), conf[k] ])
  .forEach(([ k, v ]) => v !== false ? process.env[k] = v
                                     : process.env[`NO_${k}`] = true)

// Enable cookie file by default in the same dir as the config file
if (!process.env.NO_COOKIE_FILE && !process.env.COOKIE_FILE)
  process.env.COOKIE_FILE = path.join(os.homedir(), '.spark-wallet', 'cookie')

process.env.NODE_ENV || (process.env.NODE_ENV = 'production')
process.env.VERBOSE && (process.env.DEBUG = `clightning-client,spark,superagent,${process.env.DEBUG||''}`)
process.env.ONION_PATH && (process.env.ONION = true) // --onion-path implies --onion
process.env.ONION_NONANONYMOUS && (process.env.ONION = true) // --onion-nonanonymous implies --onion
process.env.PAIRING_QR && (process.env.PRINT_QR = true) // --pairing-qr implies --print-qr
process.env.PROXY && (process.env.ALL_PROXY = process.env.PROXY) // PROXY is an alias for ALL_PROXY

if (process.env.TLS_PATH || process.env.TLS_NAME || process.env.LETSENCRYPT) process.env.FORCE_TLS = true

require('@babel/polyfill')
require('./app')
